### PR TITLE
Set the correct fieldName to get the correct value

### LINKED ISSUE #1219: FieldDescription::getValue does not allow to chain custom getter
## Feature Request

Hi,
I need to chain custom getters to add a field in a listMapper.

For instance I have 2 classes `Man` and `Dog` with the following methods :
```
Class Man {
    getDogs();

    // custom getter
    getFavoriteDog();
}
```
```
class Dog {
    getName();

    // custom getter
    getFormattedName();
}
```

`add('favoriteDog')` will work on ManAdmin and `add('formattedName')` will work on DogAdmin.
But `add('favoriteDog.formattedName')` will not display anything on ManAdmin.

From what I saw, the problem comes from FieldDescription::getValue which search only the first getter.

Does anyone have an idea to fix it without creating a new custom getter in my class `Man` ?

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Subject

I am targeting this branch, because the wrong field name is a bug, so this PR would be a bug fix.

When using the following code 
```
->add('foo.bar')
```

If the subject has a relation to a Foo entity, which has a relation to a Bar entity, the fieldDescription has the following data:
- fieldName: bar
- parentAssociationMapping: An array containing foo.

If it's an embedded entity, the fieldDescription had the following data:
- fieldName: bar
- no parentAssociationMapping.

That's why the there was extra code for embedded fields.
- in getValue
- in fixFieldDescription

But both of these code could be avoid if we correctly set the fieldDescription fieldName when looking for parentAssociationMapping. I propose to change this to:
- fieldName: foo.bar
- no parentAssociationMapping.

Then, no code is needed in fixFieldDescription. And in getValue, we just have to iterate on `explode('.', $fieldName)`.

Bonus point: it works with custom getters too.

Closes #1219.

## Changelog

```markdown
### Changed
- When using embedded fields or fake field 'foo.bar' (with custom getters in the entity), `FieldDescription::fieldName` is changed from `bar` to the correct value `foo.bar`.

### Fixed
- Support for embedded and custom getters by the FieldDescription.
```

### CODE REVIEW

**Review by wbloszyk — DISMISSED** [review]

File: src/Admin/FieldDescription.php [hunk 1]
```diff
@@ -93,19 +94,53 @@ public function isIdentifier()
     public function getValue($object)
     {
         foreach ($this->parentAssociationMappings as $parentAssociationMapping) {
-            $object = $this->getFieldValue($object, $parentAssociationMapping['fieldName']);
+            $object = $this->getChildValue($object, $parentAssociationMapping['fieldName']);
         }
 
+        /*
+         * NEXT_MAJOR: Remove this code and the related tests.
+         * The support for embedded object come from the next foreach if the fieldName is correctly set.
```
Comment (greg0ire) [inline src/Admin/FieldDescription.php:102]:
```suggestion
         * The support for embedded objects comes from the next foreach if the fieldName is correctly set.
```

File: src/Admin/FieldDescription.php [hunk 2]
```diff
@@ -93,19 +94,53 @@ public function isIdentifier()
...
+    private function getChildValue(?object $object, string $fieldName): ?object
+    {
+        if (null === $object) {
+            return null;
+        }
+
+        $child = $this->getFieldValue($object, $fieldName);
+        if (null !== $child && !\is_object($child)) {
+            throw new NoValueException(sprintf(
+                'Unexpected value when accessing to the property "%s" on the class "%s" for the field "%s".'
```
Comment (greg0ire) [inline src/Admin/FieldDescription.php:135]:
```suggestion
                'Unexpected value when accessing to property "%s" on class "%s" for field "%s".'
```

File: src/Admin/FieldDescription.php [hunk 3]
```diff
@@ -93,19 +94,53 @@ public function isIdentifier()
...
+                .' Expected object|null, got %s.',
```
Comment (greg0ire) [inline src/Admin/FieldDescription.php:136]: If you need to split the string in the code, it's probably too long in the output exception message. How about using linkebreaks in the code AND in the output?

```php
<<<'EXCEPTION'
Unexpected value when accessing to the property "%s" on the class "%s" for the field "%s".
Expected object|null, got %s.
EXCEPTION
```

**Comment by VincentLanglet** [comment]
@greg0ire I moved the support for fieldName with `.` in SonataAdmin instead: https://github.com/sonata-project/SonataAdminBundle/pull/6724

File: src/Admin/FieldDescription.php [hunk 4]
```diff
@@ -105,16 +105,6 @@ public function getValue($object)
...
 
-        $fieldMapping = $this->getFieldMapping();
-        // Support embedded object for mapping
-        // Ref: https://www.doctrine-project.org/projects/doctrine-orm/en/latest/tutorials/embeddables.html
-        if (isset($fieldMapping['declaredField'])) {
-            $parentFields = explode('.', $fieldMapping['declaredField']);
-            foreach ($parentFields as $parentField) {
-                $object = $this->getFieldValue($object, $parentField);
-            }
-        }
```
Comment (VincentLanglet) [inline src/Admin/FieldDescription.php:116]: This is not needed anymore (and will be buggy instead) if the correct fieldName is set in `ModelManager::getNewFieldDescriptionInstance`.

**Comment by VincentLanglet** [comment]
The PR seems ready to me :)

File: src/Builder/DatagridBuilder.php [hunk 5]
```diff
@@ -92,6 +92,7 @@ public function fixFieldDescription(AdminInterface $admin, FieldDescriptionInter
                     $fieldDescription->setOption('global_search', $fieldDescription->getOption('global_search', true)); // always search on string field only
                 }
 
+                // NEXT_MAJOR: Remove this, the fieldName should be correctly be set at the creation.
```
Comment (OskarStark) [inline src/Builder/DatagridBuilder.php:95]:
```suggestion
                // NEXT_MAJOR: Remove this, the fieldName should be correctly set at the creation.
```

**Review by OskarStark — APPROVED** [review]

File: src/Builder/DatagridBuilder.php (hunk 5 above)
Comment (VincentLanglet) [reply src/Builder/DatagridBuilder.php]: Fixed

**Review by OskarStark — APPROVED** [review]

**Pull request merged**